flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -841,6 +845,10 @@ protected void declineCheckpoint(long checkpointId) {
 			new CheckpointException("Task Name" + getName(), CheckpointFailureReason.CHECKPOINT_DECLINED_TASK_NOT_READY));
 	}
 
+	protected void handleCheckpointException(Exception exception) {
+		handleAsyncException(exception);
```

Comment:
Why not use existing handleAsyncException and add a message?

Reply:
This have been wrongly rebased (intermediate commit was not compiling, but the end PR was correct).

Regarding `handleAsyncException()`, I think there needs to be some refactoring/more thinking about the final exception that thrown. The doc string for `AsynchronousException` says:
```
 * An exception for wrapping exceptions that are thrown by an operator in threads other than the
 * main compute thread of that operator.
```
but moving timers and checkpoint triggers to the mailbox execution, this no longer true.